src/java.naming/share/classes/com/sun/jndi/ldap/AbstractLdapNamingEnumeration.java

```diff
@@ -251,22 +354,33 @@ private T nextImpl() throws NamingException {
             cleanup();
             throw cont.fillInException(e);
         }
+        // No fences here. nextAux() (source just below) has its own fences. The
+        // only other thing this method does is call cleanup().
+        // If nextAux() were overrideable, this method should probably have
+        // fences, but it seems OK for now without.
     }
 
     private T nextAux() throws NamingException {
-        if (posn == limit) {
-            getNextBatch();  // updates posn and limit
-        }
+        try {
+            if (posn == limit) {
+                getNextBatch();  // updates posn and limit
+            }
 
-        if (posn >= limit) {
-            cleanup();
-            throw new NoSuchElementException("invalid enumeration handle");
-        }
+            if (posn >= limit) {
+                cleanup();
+                throw new NoSuchElementException("invalid enumeration handle");
+            }
 
-        LdapEntry result = entries.elementAt(posn++);
+            LdapEntry result = entries.elementAt(posn++);
 
-        // gets and outputs DN from the entry
-        return createItem(result.DN, result.attributes, result.respCtls);
+            // gets and outputs DN from the entry
+            return createItem(result.DN, result.attributes, result.respCtls);
+        } finally {
+            // Ensure writes are visible to the Cleaner thread
+            VarHandle.fullFence();
+            // Ensure Cleaner does not run until after this method completes
+            Reference.reachabilityFence(enumCtx);
```

Comment:
Same as above: no need for memory fence and reachability fence should take 'this' as parameter.